src/synapse.ts

```diff
@@ -231,6 +241,58 @@ export class Synapse {
   getNetwork (): FilecoinNetworkType {
     return this._network
   }
+
+  /**
+   * Download a piece using discovery to find all locations
+   * Tries URLs as they're discovered, enabling fast cache hits
+   * @param commp - The piece commitment (CommP) to download
+   * @param options - Optional download options including preferred provider
+   * @returns A ReadableStream of the piece data
+   */
+  async download (
+    commp: string,
+    options?: { providerAddress?: string }
+  ): Promise<ReadableStream<Uint8Array>> {
+    const client = await this._signer.getAddress()
+    const errors: Error[] = []
+    let attemptedUrls = 0
+
+    // Try URLs as they're discovered
+    for await (const url of this._pieceDiscovery.findPiece(commp, client, options)) {
+      attemptedUrls++
+      try {
+        const response = await fetch(url)
```

Comment:
Can we modify the discovery API so that `findPiece` returns either `url` or a `Request` object, i.e. the type of the first parameter of `fetch()` API ([docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch#resource)).

This way, the discovery mechanism can provide additional metadata, e.g. request headers.

Think of a CDN that implements authorisation and requires clients to provide `Authorization: Bearer some-token`.

Reply:
yeah, this is a great idea, let's do that